### PR TITLE
Contextual evaluations for BQ

### PR DESCRIPTION
**Pull request by Ekrekr**
Context: Evaluating tables rather than just queries will perform evaluations of exactly what will run; currently the evaluation could say something is fine, when in fact it will error on something that could have been checked during compilation. This update will also allow contextual pre and post ops validation.

To make this context work, evaluate instead needs to return an array of query evaluations so that each published table from the adapter can be evaluated separately.

I also renamed `QueryEvaluationResponse` to `QueryEvaluation` because it looks neater, particularly when used in conjunction with the closed source protos. I attached the query to the proto to make handling a query evaluation more clear once it's been made.

Closed source changes still pending.

### CODE REVIEW

File: api/commands/query.ts [hunk 1]
```diff
@@ -30,11 +34,30 @@ export function run(
 
 export async function evaluate(
   dbadapter: dbadapters.IDbAdapter,
-  query: string,
-  compileConfig?: dataform.ICompileConfig
-): Promise<dataform.IQueryEvaluationResponse> {
-  const compiledQuery = await compile(query, compileConfig);
-  return await dbadapter.evaluate(compiledQuery);
+  queryStringOrTable: string | dataform.ITable,
```
Comment (BenBirt) [inline api/commands/query.ts:37]: i'm not sure this is the right place to do this.

IMO we should be moving away from `query.ts` (which is just a hacky way of injecting arbitrary SQL into the rest of a graph).

I'd expect to instead see us change the `IDbAdapter.evaluate` interface to do something like this. and when evaluating a table/assertion/operation, we'd call that interface (probably exposed somewhere higher level in `api/`), and this function would only really be used for 'selected query' evaluation.

(i would also expect the interface to change to allow all three action types, otherwise the caller has to branch on the action type which is being evaluated).

File: protos/core.proto [hunk 2]
```diff
@@ -186,13 +186,14 @@ message QueryEvaluationError {
...
 
-message QueryEvaluationResponse {
+message QueryEvaluation {
   enum QueryEvaluationStatus {
     SUCCESS = 0;
     FAILURE = 1;
   }
   QueryEvaluationStatus status = 1;
   QueryEvaluationError error = 2;
+  string query = 3;
```
Comment (BenBirt) [inline protos/core.proto:196]: i'm not super clear on why this is useful? i would expect callers to `evaluate` single queries/tables/assertions/operations individually.

Comment (Ekrekr) [reply protos/core.proto]: This will be needed to make line numbers match in the closed source UI I believe. As the statement that is run against the warehouse will only be fully built in the backend (when the `evaluate` part of `@dataform/api` runs `publishTasks`), this will need to be send back.

**Comment by Ekrekr** [comment]
Have made PR specific to BQ; as the API's `evaluate` isn't actually used anywhere currently (correct me if I'm wrong), it's OK to make a dramatic change to it.

File: components/dialog.tsx [hunk 3]
```diff
@@ -14,6 +14,7 @@ export const Dialog = ({
     <BlueprintDialog
       className={styles.dialog}
       isCloseButtonShown={false}
+      title=""
```
Comment (Ekrekr) [inline components/dialog.tsx:17]: This prevents a console warning that I was getting spammed by; can move in to separate PR if you want.

File: core/adapters/bigquery.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ export class BigQueryAdapter extends Adapter implements IAdapter {
   public publishTasks(
     table: dataform.ITable,
     runConfig: dataform.IRunConfig,
-    tableMetadata: dataform.ITableMetadata
+    tableMetadata?: dataform.ITableMetadata
```
Comment (Ekrekr) [inline core/adapters/bigquery.ts:19]: There's a slight issue with metadata; in order for the evaluation to be 100% accurate of evaluating what will run, metadata is required. This is annoying because acquiring metadata requires querying the entire warehouse (?), which obviously we don't want to happen every time a file is edited.

File: tests/integration/bigquery.spec.ts [hunk 5]
```diff
@@ -308,6 +308,87 @@ suite("@dataform/integration/bigquery", ({ before, after }) => {
...
+            name: "example_incremental",
+            database: "dataform-integration-tests"
+          }
+        }),
+        { useSingleQueryPerAction: true }
+      );
+      expect(evaluation.status).to.equal(dataform.QueryEvaluation.QueryEvaluationStatus.SUCCESS);
+    });
+  });
+
```
Comment (Ekrekr) [inline tests/integration/bigquery.spec.ts:391]: I haven't added a test for incremental yet

File: api/commands/build.ts [hunk 6]
```diff
@@ -62,7 +62,7 @@ export class Builder {
       useSingleQueryPerAction:
         !this.prunedGraph.projectConfig?.hasOwnProperty("useSingleQueryPerAction") ||
         typeof this.prunedGraph.projectConfig?.useSingleQueryPerAction === "undefined"
-          ? this.prunedGraph.projectConfig.useSingleQueryPerAction
+          ? false
```
Comment (BenBirt) [inline api/commands/build.ts:100]: please merge with master, i made some fixes to this code recently

File: api/dbadapters/index.ts [hunk 7]
```diff
@@ -26,7 +26,10 @@ export interface IDbAdapter {
       maxResults?: number;
     }
   ): Promise<IExecutionResult>;
-  evaluate(statement: string): Promise<dataform.IQueryEvaluationResponse>;
+  evaluate(
+    queryOrTable: string | dataform.ITable | dataform.IOperation | dataform.IAssertion,
```
Comment (BenBirt) [inline api/dbadapters/index.ts:30]:
```suggestion
    queryOrAction: string | dataform.ITable | dataform.IOperation | dataform.IAssertion,
```

File: api/dbadapters/redshift.ts [hunk 8]
```diff
@@ -68,12 +68,12 @@ export class RedshiftDbAdapter implements IDbAdapter {
     const statementWithExplain = `explain ${statement}`;
```
Comment (BenBirt) [inline api/dbadapters/redshift.ts:81]: i'm not clear on how this code can still be compiling. you should need to change `evaluate` to match the new interface (and then evaluate the query(s) according to the type that is passed in)

File: api/dbadapters/bigquery.ts [hunk 9]
```diff
@@ -91,22 +93,57 @@ export class BigQueryDbAdapter implements IDbAdapter {
...
-        status: dataform.QueryEvaluationResponse.QueryEvaluationStatus.SUCCESS
-      });
-    } catch (e) {
-      return dataform.QueryEvaluationResponse.create({
-        status: dataform.QueryEvaluationResponse.QueryEvaluationStatus.FAILURE,
-        error: parseBigqueryEvalError(e)
-      });
+  public async evaluate(
+    query: string | dataform.Table | dataform.Operation | dataform.Assertion,
+    projectConfig?: dataform.IProjectConfig
```
Comment (BenBirt) [inline api/dbadapters/bigquery.ts:103]: AFAICT this is actually required now, otherwise the code won't work? seems like it should be required here and in the interface

File: api/dbadapters/bigquery.ts [hunk 10]
```diff
@@ -91,22 +93,57 @@ export class BigQueryDbAdapter implements IDbAdapter {
...
+        if (query instanceof dataform.Table) {
+          executionTasks = coreAdapter
+            .publishTasks(
+              query,
+              { useSingleQueryPerAction: projectConfig?.useSingleQueryPerAction },
+              null
+            )
+            .build();
+        } else if (query instanceof dataform.Operation) {
+          executionTasks = query.queries.map(statement =>
```
Comment (BenBirt) [inline api/dbadapters/bigquery.ts:120]: OK, for this case i can see that we're going to need to return a list of evaluation responses, I'm fine with that API change!

File: tests/api/api.spec.ts [hunk 11]
```diff
@@ -701,14 +701,6 @@ suite("@dataform/api", () => {
       {
         warehouse: "bigquery",
         expectedQuery: "preOps;create or replace table `database.schema.b` as query;postOps"
-      },
-      {
-        warehouse: "sqldatawarehouse",
```
Comment (BenBirt) [inline tests/api/api.spec.ts:731]: does this test case need removing?

File: tests/integration/bigquery.spec.ts [hunk 12]
```diff
@@ -346,6 +346,111 @@ suite("@dataform/integration/bigquery", { parallel: true }, ({ before, after })
     ]);
   });
 
+  suite("evaluate", async () => {
```
Comment (BenBirt) [inline tests/integration/bigquery.spec.ts:349]: i suggest making this suite `parallel`

File: tests/integration/bigquery_project/dataform.json [hunk 13]
```diff
@@ -2,5 +2,7 @@
   "warehouse": "bigquery",
   "defaultDatabase": "dataform-integration-tests",
   "defaultSchema": "df_integration_test",
-  "assertionSchema": "df_integration_test_assertions"
+  "assertionSchema": "df_integration_test_assertions",
+  "useRunCache": true,
```
Comment (BenBirt) [inline tests/integration/bigquery_project/dataform.json:6]: please undo this

File: tests/integration/bigquery_project/dataform.json [hunk 14]
```diff
@@ -2,5 +2,7 @@
...
+  "useSingleQueryPerAction": true
```
Comment (BenBirt) [inline tests/integration/bigquery_project/dataform.json:7]: IMO i'd just set this in the tests themselves; that would be consistent with run caching

File: api/dbadapters/bigquery.ts (hunk 9 above)
Comment (Ekrekr) [reply api/dbadapters/bigquery.ts:103]: Yeah pretty weird that that's allowed. Have updated to use new interface in other adapters, and added a throw if they're given things other than strings.

File: api/dbadapters/redshift.ts (hunk 8 above)
Comment (Ekrekr) [reply api/dbadapters/redshift.ts:81]: Yeah pretty weird that that's allowed. Have updated to use new interface in other adapters, and added a throw if they're given things other than strings.

File: tests/integration/bigquery.spec.ts (hunk 12 above)
Comment (Ekrekr) [reply tests/integration/bigquery.spec.ts:349]: Ooh

File: tests/integration/bigquery_project/dataform.json (hunk 14 above)
Comment (Ekrekr) [reply tests/integration/bigquery_project/dataform.json]: Nice

File: api/dbadapters/redshift.ts (hunk 8 above)
Comment (Ekrekr) [reply api/dbadapters/redshift.ts]: With TODOs

File: api/dbadapters/bigquery.ts [hunk 15]
```diff
@@ -96,22 +98,57 @@ export class BigQueryDbAdapter implements IDbAdapter {
...
+        if (queryOrAction instanceof dataform.Table) {
+          executionTasks = coreAdapter
+            .publishTasks(
+              queryOrAction,
+              { useSingleQueryPerAction: projectConfig?.useSingleQueryPerAction },
+              null
+            )
+            .build();
+        } else if (queryOrAction instanceof dataform.Operation) {
+          executionTasks = queryOrAction.queries.map(statement =>
```
Comment (BenBirt) [inline api/dbadapters/bigquery.ts:120]: i think the last comment maybe was confusing? i mean that we should change (back to what you had originally) evaluate to return an array, so that cases like this make sense.

File: api/dbadapters/bigquery.ts [hunk 16]
```diff
@@ -96,22 +98,57 @@ export class BigQueryDbAdapter implements IDbAdapter {
...
+  ) {
+    let executionTasks: dataform.ExecutionTask[];
+    if (typeof queryOrAction === "string") {
+      executionTasks = [dataform.ExecutionTask.create({ statement: queryOrAction })];
+    } else {
+      try {
+        const coreAdapter = new BigQueryAdapter(projectConfig, version);
+        if (queryOrAction instanceof dataform.Table) {
+          executionTasks = coreAdapter
+            .publishTasks(
```
Comment (BenBirt) [inline api/dbadapters/bigquery.ts:113]: just pinged you on chat, but this changes the semantics of evaluation a bit (sorry for just noticing this part).

i don't think we want to evaluate what is effectively an `ExecutionAction`, at least not in that part of the UI. I think a solution to that should be separate

File: api/dbadapters/bigquery.ts (hunk 15 above)
Comment (Ekrekr) [reply api/dbadapters/bigquery.ts]: Sure, have updated
